README.md
# HOG features (Histogram of oriented gradients)

  [![build status][travis-image]][travis-url]
  [![npm download][download-image]][download-url]

## Principe

The main feature of this repository will compute the [HOG features](https://en.wikipedia.org/wiki/Histogram_of_oriented_gradients) of an image. The HOG features (called HOG descriptor too) are useful for image recognition and image detection. You can find a good tutorial about HOG features [here](http://mccormickml.com/2013/05/09/hog-person-detector-tutorial/).


## Usage

### extractHOG(image[, options])

Generate a vector which corresponds to the HOG descriptor of an image.
Returns an array of float.

__arguments__

* `image` - an Image
* `options` - an optional object

__options__

* `cellSize`: length of cell in px (default: 4).
* `blockSize`: length of block in number of cells (default: 2).
* `blockStride`: number of cells to slide block window by (default: block-size / 2).
* `bins`: bins per histogram (default: 6).
* `norm`: norm block normalization method (default: "L2". Other possibilities : "L1" and "L1-sqrt").

## Example

```js
'use strict';

const {Image} = require('image-js');
const hog = require('hog-features');

var file = __dirname + '/__test__/beachball.png';

Image.load(file).then(function (image) {
    var descriptor = hog.extractHOG(image);
    console.log(descriptor);
});
```

## Tutorial

You can find a tutorial where the HOG features is used with an SVM classifier to classify road signs. [Here is the tutorial](https://github.com/jajoe/tutorial-nodejs/tree/master/image-classification).

## License

[MIT](./LICENSE)

Inspired by [harthur implementation](https://github.com/harthur/hog-descriptor)

[travis-image]: https://img.shields.io/travis/image-js/hog/master.svg?style=flat-square
[travis-url]: https://travis-ci.org/image-js/hog
[download-image]: https://img.shields.io/npm/dm/hog-features.svg?style=flat-square
[download-url]: https://www.npmjs.com/package/hog-features


History.md
<a name="0.2.0"></a>
# [0.2.0](https://github.com/image-js/hog/compare/v0.1.0...v0.2.0) (2018-03-23)



<a name="0.1.0"></a>
# [0.1.0](https://github.com/image-js/hog/compare/v0.0.3...v0.1.0) (2017-12-11)



<a name="0.0.3"></a>
## [0.0.3](https://github.com/image-js/hog/compare/v0.0.2...v0.0.3) (2017-07-20)



<a name="0.0.2"></a>
## [0.0.2](https://github.com/image-js/hog/compare/v0.0.1...v0.0.2) (2017-07-20)



<a name="0.0.1"></a>
## 0.0.1 (2017-07-20)





test/test.js
var {default: Image} = require('image-js');
var HOG = require('../src');

describe('hog size test', () => {
    it('Main test', async () => {
        const promises = [Image.load(__dirname + "/beachball.jpg")];
        const sizeImage = 110; // image test1/png is square sizeImage*sizeImage
        const images = await Promise.all(promises);
        const options = {
            cellSize: 4,
            bins: 6,
            blockSize: 2,
            blockStride: 1
        };
        const descriptorSize = HOG.extractHOG(images[0], options).length;
        const sizeExpected = (Math.floor((parseFloat(sizeImage)/options.cellSize - options.blockSize)/(options.blockSize - options.blockStride)) + 1)**2*options.bins*options.cellSize;
        expect(descriptorSize).toEqual(sizeExpected);
    });
});


__tests__/test.js
'use strict';

const path = require('path');
const {Image} = require('image-js');

const HOG = require('../src');

describe('hog size test', () => {
    it('Main test', async () => {
        const image = await Image.load(path.join(__dirname, 'beachball.png'));
        const sizeImage = 110; // image test1/png is square sizeImage*sizeImage
        const options = {
            cellSize: 4,
            bins: 6,
            blockSize: 2,
            blockStride: 1
        };
        const descriptorSize = HOG.extractHOG(image, options).length;
        const sizeExpected = (Math.floor((parseFloat(sizeImage) / options.cellSize - options.blockSize) / (options.blockSize - options.blockStride)) + 1) ** 2 * options.bins * options.cellSize;
        expect(descriptorSize).toEqual(sizeExpected);
    });
});


test/test-main.js
var {default: Image} = require('image-js');
var hog = require("../src");

var file = __dirname + "/beachball.jpg";

Image.load(file).then(function(image){
  var descriptor = hog.extractHOG(image);
  console.log(descriptor);
});



hog.js
var processing = require("./processingHog"),
    distance = require("ml-distance");

const PI_RAD = 180 / Math.PI;

module.exports = {
  extractHOG: extractHOG,
  extractHistograms: extractHistograms
}

// also export all the functions from processing.js
for (var func in processing) {
  module.exports[func] = processing[func];
}

/**
 * Extract the HOG of an image
 * @param {image} Image - image to transform into a HOG descriptor
 * @param {object} options
 * @return {Array} Array with the value of the HOG descriptor
 */

function extractHOG(image, options = {}) {
  var blockSize = options.blockSize || 2;
  var blockStride = options.blockStride || (blockSize / 2);
  var norm = options.norm || "L2";

  var histograms = extractHistograms(image, options);

  var blocks = [];
  var blocksHigh = histograms.length - blockSize + 1;
  var blocksWide = histograms[0].length - blockSize + 1;

  for (var y = 0; y < blocksHigh; y += blockStride) {
    for (var x = 0; x < blocksWide; x += blockStride) {
      var block = getBlock(histograms, x, y, blockSize);
      normalize(block, norm);
      blocks.push(block);
    }
  }
  return Array.prototype.concat.apply([], blocks);
}

/**
 * Extract the histogram from an image
 * @param {image} Image - image to transform into a HOG descriptor
 * @param {object} options
 * @return {Array 2D} Array 2D with the histogram, based on the gradient vectors
 */

function extractHistograms(image, options = {}) {
  var vectors = processing.gradientVectors(image);

  var cellSize = options.cellSize || 4;
  var bins = options.bins || 6;

  var cellsWide = Math.floor(vectors[0].length / cellSize);
  var cellsHigh = Math.floor(vectors.length / cellSize);

  var histograms = new Array(cellsHigh);

  for (var i = 0; i < cellsHigh; i++) {
    histograms[i] = new Array(cellsWide);

    for (var j = 0; j < cellsWide; j++) {
      histograms[i][j] = getHistogram(vectors, j * cellSize, i * cellSize,
                                      cellSize, bins);
    }
  }
  return histograms;
}

/**
 * Extract a sqare block from a matrix
 * @param {Array 2D} Matrix
 * @param {number} x
 * @param {number} y
 * @param {number} length
 * @return {Array 2D} square block extracted from the matrix
 */

function getBlock(matrix, x, y, length) {
  var square = [];
  for (var i = y; i < y + length; i++) {
    for (var j = x; j < x + length; j++) {
      square.push(matrix[i][j]);
    }
  }
  return Array.prototype.concat.apply([], square);
}

function getHistogram(elements, x, y, size, bins) {
  var histogram = zeros(bins);

  for (var i = 0; i < size; i++) {
    for (var j = 0; j < size; j++) {
      var vector = elements[y + i][x + j];
      var bin = binFor(vector.orient, bins);
      histogram[bin] += vector.mag;
    }
  }
  return histogram;
}

function binFor(radians, bins) {
  var angle = radians * (PI_RAD);
  if (angle < 0) {
    angle += 180;
  }

  // center the first bin around 0
  angle += 90 / bins;
  angle %= 180;

  var bin = Math.floor(angle / 180 * bins);
  return bin;
}

function zeros(size) {
  var array = new Array(size);
  for (var i = 0; i < size; i++) {
    array[i] = 0;
  }
  return array;
}

function normalize(vector, norm){
  var epsilon = 0.00001;
  if(norm === "L1"){
    var norm = 0;
    for (var i = 0; i < vector.length; i++) {
      norm += Math.abs(vector[i]);
    }
    var denom = norm + epsilon;

    for (var i = 0; i < vector.length; i++) {
      vector[i] /= denom;
    }
  }
  else if(norm === "L1-sqrt"){
    var norm = 0;
    for (var i = 0; i < vector.length; i++) {
      norm += Math.abs(vector[i]);
    }
    var denom = norm + epsilon;

    for (var i = 0; i < vector.length; i++) {
      vector[i] = Math.sqrt(vector[i] / denom);
    }
  }
  else{ // i.e norm === "L2"
    var sum = 0;
    for (var i = 0; i < vector.length; i++) {
      sum += Math.pow(vector[i], 2);
    }
    var denom = Math.sqrt(sum + epsilon);
    for (var i = 0; i < vector.length; i++) {
      vector[i] /= denom;
    }
  }
}


src/processingHog.js
'use strict';

function intensities(image) {
    if (image.components === 1) {
        return image;
    } else {
        return image.grey({algorithm: 'luma601'});
    }
}

function gradients(imagedata) {
    return _gradients(intensities(imagedata));
}

function _gradients(intensities) {
    const height = intensities.height;
    const width = intensities.width;
    const maxValue = intensities.maxValue;

    const gradX = new Array(height);
    const gradY = new Array(height);

    for (var y = 0; y < height; y++) {
        gradX[y] = new Array(width);
        gradY[y] = new Array(width);

        for (var x = 0; x < width; x++) {
            var prevX = x === 0 ? 0 : intensities.getValueXY(x - 1, y, 1) / maxValue;
            var nextX = x === width - 1 ? 0 : intensities.getValueXY(x + 1, y, 1) / maxValue;
            var prevY = y === 0 ? 0 : intensities.getValueXY(x, y - 1, 1) / maxValue;
            var nextY = y === height - 1 ? 0 : intensities.getValueXY(x, y + 1, 1) / maxValue;

            // kernel [-1, 0, 1]
            gradX[y][x] = -prevX + nextX;
            gradY[y][x] = -prevY + nextY;
        }
    }

    return {
        x: gradX,
        y: gradY
    };
}

function gradientVectors(image) {
    return _gradientVectors(intensities(image));
}

function _gradientVectors(intensities) {
    const height = intensities.height;
    const width = intensities.width;
    const maxValue = intensities.maxValue;

    const vectors = new Array(height);

    for (var y = 0; y < height; y++) {
        vectors[y] = new Array(width);
        for (var x = 0; x < width; x++) {
            var prevX = x === 0 ? 0 : intensities.getValueXY(x - 1, y, 0) / maxValue;
            var nextX = x === width - 1 ? 0 : intensities.getValueXY(x + 1, y, 0) / maxValue;
            var prevY = y === 0 ? 0 : intensities.getValueXY(x, y - 1, 0) / maxValue;
            var nextY = y === height - 1 ? 0 : intensities.getValueXY(x, y + 1, 0) / maxValue;

            // kernel [-1, 0, 1]
            var gradX = -prevX + nextX;
            var gradY = -prevY + nextY;

            vectors[y][x] = {
                mag: Math.sqrt(Math.pow(gradX, 2) + Math.pow(gradY, 2)),
                orient: Math.atan2(gradY, gradX)
            };
        }
    }
    return vectors;
}

module.exports = {
    intensities,
    gradients,
    gradientVectors
};


src/index.js
'use strict';

const processing = require('./processingHog');

const PI_RAD = 180 / Math.PI;

module.exports = {
    extractHOG: extractHOG,
    extractHistograms: extractHistograms
};

// also export all the functions from processing.js
for (var func in processing) {
    module.exports[func] = processing[func];
}

/**
 * Extract the HOG of an image
 * @param {image} image - image to transform into a HOG descriptor
 * @param {object} options
 * @return {Array} Array with the value of the HOG descriptor
 */

function extractHOG(image, options = {}) {
    const {
        blockSize = 2,
        blockStride = blockSize / 2,
        norm = 'L2'
    } = options;

    var histograms = extractHistograms(image, options);

    var blocks = [];
    var blocksHigh = histograms.length - blockSize + 1;
    var blocksWide = histograms[0].length - blockSize + 1;

    for (var y = 0; y < blocksHigh; y += blockStride) {
        for (var x = 0; x < blocksWide; x += blockStride) {
            var block = getBlock(histograms, x, y, blockSize);
            normalize(block, norm);
            blocks.push(block);
        }
    }
    return Array.prototype.concat.apply([], blocks);
}

/**
 * Extract the histogram from an image
 * @param {image} image - image to transform into a HOG descriptor
 * @param {object} options
 * @return {Array<Array<number>>} Array 2D with the histogram, based on the gradient vectors
 */

function extractHistograms(image, options = {}) {
    var vectors = processing.gradientVectors(image);

    var cellSize = options.cellSize || 4;
    var bins = options.bins || 6;

    var cellsWide = Math.floor(vectors[0].length / cellSize);
    var cellsHigh = Math.floor(vectors.length / cellSize);

    var histograms = new Array(cellsHigh);

    for (var i = 0; i < cellsHigh; i++) {
        histograms[i] = new Array(cellsWide);

        for (var j = 0; j < cellsWide; j++) {
            histograms[i][j] = getHistogram(vectors, j * cellSize, i * cellSize,
                cellSize, bins);
        }
    }
    return histograms;
}

/**
 * Extract a sqare block from a matrix
 * @param {Array<Array<number>>} matrix
 * @param {number} x
 * @param {number} y
 * @param {number} length
 * @return {Array<Array<number>>} square block extracted from the matrix
 */

function getBlock(matrix, x, y, length) {
    var square = [];
    for (var i = y; i < y + length; i++) {
        for (var j = x; j < x + length; j++) {
            square.push(matrix[i][j]);
        }
    }
    return Array.prototype.concat.apply([], square);
}

/**
 * Extract the histogram of a part of the image (a cell with coordinate x and y)
 * @param {Array<Array<number>>} elements - gradient vectors of the image
 * @param {number} x
 * @param {number} y
 * @param {number} size - cellSize
 * @param {number} bins - number of bins per histogram
 * @return {Array<number>} Array 1D with the histogram of the cell, based on the gradient vectors
 */

function getHistogram(elements, x, y, size, bins) {
    var histogram = new Array(bins).fill(0);

    for (var i = 0; i < size; i++) {
        for (var j = 0; j < size; j++) {
            var vector = elements[y + i][x + j];
            var bin = binFor(vector.orient, bins);
            histogram[bin] += vector.mag;
        }
    }
    return histogram;
}

function binFor(radians, bins) {
    var angle = radians * (PI_RAD);
    if (angle < 0) {
        angle += 180;
    }

    // center the first bin around 0
    angle += 90 / bins;
    angle %= 180;

    return Math.floor(angle / 180 * bins);
}

/**
 * Normalize a vector given with one of these norms : L1, L1-sqrt or L2 (norm by default). No return value, the input vector is modified.
 * @param {Array<number>} vector
 * @param {string} norm - should be "L1", "L1-sqrt" or "L2". Else, the norm will be the norm L2.
 */

function normalize(vector, norm) {
    var epsilon = 0.00001;
    var sum, denom, i;
    if (norm === 'L1') {
        sum = 0;
        for (i = 0; i < vector.length; i++) {
            sum += Math.abs(vector[i]);
        }
        denom = sum + epsilon;

        for (i = 0; i < vector.length; i++) {
            vector[i] /= denom;
        }
    } else if (norm === 'L1-sqrt') {
        sum = 0;
        for (i = 0; i < vector.length; i++) {
            sum += Math.abs(vector[i]);
        }
        denom = sum + epsilon;

        for (i = 0; i < vector.length; i++) {
            vector[i] = Math.sqrt(vector[i] / denom);
        }
    } else { // i.e norm === "L2"
        sum = 0;
        for (i = 0; i < vector.length; i++) {
            sum += vector[i] * vector[i];
        }
        denom = Math.sqrt(sum + epsilon);
        for (i = 0; i < vector.length; i++) {
            vector[i] /= denom;
        }
    }
}
